### DISCUSSION TITLE
[RFC] New Component: Tag.Group

### DISCUSSION THREAD

**Discussion by BoyYangzai** [ant-design/ant-design#42911] [Ideas]
![image](https://github.com/ant-design/ant-design/assets/94534613/2790a6e6-7345-4f93-9c5a-80a579f20654)
# Summary：
Tag.Group 组件是一个包装组件，它接受一个 Tag 组件的数组并将它们组成一个组。如果标签过多，组件将折叠剩余的标签，并显示折叠的标签的工具提示。

# Motivation：
某天，在字节前端大群看到有人在咨询这个功能，很多人都说自己曾经做过类似的需求，但是很难有一个统一的方案<br>现将使用 `rc-overflow` 和 `Tag` 封装一层组成一个新的更便捷的组件在 `antd`中，快速完成这个需求

<h2 style="padding: 0px; margin: 0px; color: rgb(5, 14, 23); font-family: -apple-system, system-ui, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen-Sans, Ubuntu, Cantarell, &quot;Helvetica Neue&quot;, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(241, 242, 242); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">API</h3>

名称 | 类型 | 默认值 | 描述
-- | -- | -- | --
data | TagsGroupData | - | 要显示的标签数据。
restTagProps | TagProps | - | 自定义折叠标签的Props。
closeTooltip | boolean | false | 是否在工具提示中显示关闭图标。
resetTagToolTipConfig | TooltipProps | - | 折叠标签的工具提示属性。
renderRestPlacement | (args: TagsGroupData) => React.ReactNode | - | 自定义折叠标签占位符。
renderRawItem | (omittedItems: TagsGroupDataItem) => React.ReactElement | - | 自定义显示标签 同 rc-overflow。
renderRawRest | (omittedItems: TagsGroupData) => React.ReactElement | - | 自定义折叠标签  同 rc-overflow。
prefixCls | string | 'ant' | CSS 类名前缀。
### TS Type：
```ts
export type TagsGroupDataItem = TagProps & {
  label?: ((args: TagsGroupDataItem | TagsGroupData) => ReactNode) | string;
  key?: any;
  onClose: (newData: TagsGroupData) => void;
};

export type TagsGroupData = TagsGroupDataItem[];
```

<h3 style="padding: 0px; margin: 0px; color: rgb(5, 14, 23); font-family: -apple-system, system-ui, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen-Sans, Ubuntu, Cantarell, &quot;Helvetica Neue&quot;, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(241, 242, 242); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">注意事项</h3><ul style="list-style: disc; padding: 0px 0px 0px 2em; margin: 0px 0px 1.2em; color: rgb(5, 14, 23); font-family: -apple-system, system-ui, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen-Sans, Ubuntu, Cantarell, &quot;Helvetica Neue&quot;, sans-serif; font-size: 18px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(241, 242, 242); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="padding: 0px; margin: 0px;"><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace;">data</code><span> </span>属性是必需的，否则组件将不会渲染任何标签。</li><li style="padding: 0px; margin: 0px;">如果<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">restTagProps</code><span> </span>属性为<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">undefined</code>，则默认折叠标签将显示为“+n...”文本，其中<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace;">n</code><span> </span>是剩余标签的数量。</li><li style="padding: 0px; margin: 0px;">如果<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">renderRestPlacement</code><span> </span>属性为<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">undefined</code>，则默认折叠标签将显示为“+n...”文本，其中<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace;">n</code><span> </span>是剩余标签的数量。</li><li style="padding: 0px; margin: 0px;">如果<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">renderRawItem</code><span> </span>属性为<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">undefined</code>，则默认标签将显示为<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace;">Tag</code><span> </span>组件。</li><li style="padding: 0px; margin: 0px;">如果<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">renderRawRest</code><span> </span>属性为<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">undefined</code>，则默认折叠标签将显示为<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace; margin-bottom: var(--spacing-medium);">Tag</code><span> </span>组件，并应用<span> </span><code node="[object Object]" style="background-color: transparent; white-space: pre-wrap; padding: 0px; font-family: Consolas, monaco, monospace;">restTagProps</code><span> </span>属性。</li></ul>

# First Version 
```ts
import classNames from 'classnames';
import RcOverflow from 'rc-overflow';
import React, { ReactNode, useMemo } from 'react';

import { ConfigContext } from '../config-provider';
import type { TooltipProps } from '../tooltip';
import Tooltip from '../tooltip';
import type { TagProps } from './';
import Tag from './index';
import useStyle from './style';

export type TagsGroupDataItem = TagProps & {
  label?: ((args: TagsGroupDataItem | TagsGroupData) => ReactNode) | string;
  key?: any;
  onClose: (newData: TagsGroupData) => void;
};

export type TagsGroupData = TagsGroupDataItem[];

export interface GroupProps {
  data: TagsGroupData;
  restTagProps?: TagProps;
  closeTooltip?: boolean;
  resetTagToolTipConfig?: TooltipProps;
  renderRestPlacement?: (args: TagsGroupData) => ReactNode;
  renderRawItem?: (omittedItems: TagsGroupDataItem) => React.ReactElement;
  renderRawRest?: (omittedItems: TagsGroupData) => React.ReactElement;
  prefixCls?: string;
}

const Group: React.FC<GroupProps> = ({
  data,
  restTagProps,
  closeTooltip,
  resetTagToolTipConfig,
  renderRestPlacement,
  renderRawItem,
  renderRawRest,
  prefixCls: customizePrefixCls,
}) => {
  const renderRest = (restTags: TagsGroupData) => {
    const mergedCloseTooltip = useMemo(() => {
      return !closeTooltip && restTags.length > 0;
    }, [closeTooltip, restTags.length]);

    const renderToolTip = () =>
      restTags.map((tag) => {
        const { onClose, label, ...restTags } = tag;
        return (
          <Tag {...restTags} onClose={() => onClose?.(data)}>
            {typeof label === 'function' ? label(tag) : label}
          </Tag>
        );
      });

    const renderMergedRestPlacement = () => {
      return renderRestPlacement ? (
        renderRestPlacement(restTags)
      ) : (
        <Tag style={{ cursor: 'pointer' }} {...restTagProps}>
          +{restTags.length}...
        </Tag>
      );
    };
    return (
      <>
        {mergedCloseTooltip ? (
          <Tooltip
            title={renderToolTip}
            color="white"
            overlayStyle={{ width: '200px' }}
            {...resetTagToolTipConfig}
          >
            {renderMergedRestPlacement()}
          </Tooltip>
        ) : (
          renderMergedRestPlacement()
        )}
      </>
    );
  };

  const renderTag = (tag: TagsGroupDataItem) => {
    const { onClose, label, ...restProps } = tag;
    return (
      <Tag {...restProps} onClose={() => onClose?.(data)}>
        {typeof label === 'function' ? label(tag) : label}
      </Tag>
    );
  };

  // Style
  const { getPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = getPrefixCls('tags-group', customizePrefixCls);
  const [wrapSSR, hashId] = useStyle(`${prefixCls}`);

  const cls = classNames(hashId);

  return wrapSSR(
    <RcOverflow
      data={[...data!]}
      renderItem={renderTag}
      renderRest={renderRest}
      renderRawRest={renderRawRest}
      renderRawItem={renderRawItem}
      maxCount={'responsive'}
      className={cls}
      prefixCls={`${prefixCls}-overflow`}
    ></RcOverflow>,
  );
};

export default Group;
```

# Basic Example Usage
```ts
import { DownOutlined, SyncOutlined } from '@ant-design/icons';
import { Divider, Tag } from 'antd';
import React from 'react';
import type { TagsGroupData } from '../Group';

function randomColor() {
  const r = Math.floor(Math.random() * 256);
  const g = Math.floor(Math.random() * 256);
  const b = Math.floor(Math.random() * 256);
  const hex = '#' + ((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1);
  return hex;
}

const App: React.FC = () => {
  const [data, setData] = React.useState(createData(12));

  function createData(count: number) {
    const generatedData: TagsGroupData = new Array(count).fill(1).map((item, index) => ({
      color: randomColor(),
      label: () => (
        <span>
          <SyncOutlined spin />
          {`Tag ${index}`}
        </span>
      ),
      closable: true,
      key: index,
      onClose(newData: TagsGroupData) {
        const newTags = newData.filter((newTag) => newTag.key !== index);
        setData([...newTags]);
      },
    }));
    return generatedData;
  }

  return (
    <>
      <Tag.Group data={data} restTagProps={{ color: 'blue' }}></Tag.Group>
      <Divider />
      <Tag.Group
        data={data}
        renderRestPlacement={() => <DownOutlined style={{ fontSize: '10px', cursor: 'pointer' }} />}
      ></Tag.Group>
    </>
  );
};

export default App;
```
# Effect show
![image](https://github.com/ant-design/ant-design/assets/94534613/94604768-a0c3-4bc4-b88d-ddfceba38504)
# Related Link
- https://github.com/ant-design/ant-design/pull/42912

**Comment by afc163** [comment]
@zombieJ @MadCcc 看看，API 应该是要好好改改，现在这一版命名风格和 antd 风格不太同意，另外 renderXxx 太多了。
